File: bridge.py
'''
n people wish to cross a bridge at night. A group of at most two people may cross at any time, and each group must have a flashlight. Only one flashlight is available among the n people, so some sort of shuttle arrangement must be arranged in order to return the flashlight so that more people may cross.
Each person has a different crossing speed; the speed of a group is determined by the speed of the slower member. Your job is to determine a strategy that gets all n people across the bridge in the minimum time.
'''

peopleCount = int(input())

people = []

for i in range(peopleCount):
    people.append(int(input()))

people.sort()

print("list of people: " + str(people))

def updateSlowest(people):
    slowest = people[0]
    for i in people:
        if i > slowest:
            slowest = i
    return slowest

def updateSecondSlowest(people):
    slowest = updateSlowest(people)
    secondSlowest = people[0]
    for i in people:
        if i > secondSlowest and i < slowest:
            secondSlowest = i
    return secondSlowest

fastest = people[0]
secondFastest = people[1]

people_left = people
totalTime = 0
sequence = []

while len(people_left) > 3:
    slowest = updateSlowest(people_left)
    secondSlowest = updateSecondSlowest(people_left)

    firstMethod = 2 * fastest + slowest + secondSlowest
    secondMethod = 2 * slowest + fastest + secondFastest

    if firstMethod < secondMethod:
        totalTime += firstMethod
        sequence.append([fastest, secondFastest])
        sequence.append([fastest])
        sequence.append([slowest, secondSlowest])
        sequence.append([secondFastest])
    else:
        totalTime += secondMethod
        sequence.append([fastest, slowest])
        sequence.append([fastest])
        sequence.append([fastest, secondFastest])
        sequence.append([fastest])
    
    people_left.remove(slowest)
    people_left.remove(secondSlowest)

if len(people_left) == 3:
    sequence.append([people_left[0], people_left[1]])
    sequence.append([people_left[0]])
    a = [people_left[0], people_left[2]]
    sequence.append(a)
    people_left.clear()



if len(people_left) == 2:
    sequence.append(people_left)
    people_left.clear()


if len(people_left) == 1:
    sequence.append(people_left)
    people_left.clear()

print(totalTime)

# print every element in the list on a new line while also removing the brackets and commas
for i in sequence:
    print(str(i).replace("[", "").replace("]", "").replace(",", ""))

File: bridge.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>

using namespace std;

int updateSlowest(vector<int> people) {
    int slowest = people[0];
    for (int i : people) {
        if (i > slowest) {
            slowest = i;
        }
    }
    return slowest;
}

int updateSecondSlowest(vector<int> people) {
    int slowest = updateSlowest(people);
    int secondSlowest = people[0];
    for (int i : people) {
        if (i > secondSlowest && i < slowest) {
            secondSlowest = i;
        }
    }
    return secondSlowest;
}

int main() {
    int peopleCount;
    cin >> peopleCount;
    vector<int> people;
    for (int i = 0; i < peopleCount; i++) {
        int temp;
        cin >> temp;
        people.push_back(temp);
    }
    sort(people.begin(), people.end());

    int fastest = people[0];
    int secondFastest = people[1];

    vector<int> people_left = people;
    int totalTime = 0;
    vector<vector<int>> sequence;

    while (people_left.size() > 3) {
        int slowest = updateSlowest(people_left);
        int secondSlowest = updateSecondSlowest(people_left);

        int firstMethod = 2 * fastest + slowest + secondSlowest;
        int secondMethod = 2 * slowest + fastest + secondFastest;

        if (firstMethod < secondMethod) {
            totalTime += firstMethod;
            sequence.push_back({fastest, secondFastest});
            sequence.push_back({fastest});
            sequence.push_back({slowest, secondSlowest});
            sequence.push_back({secondFastest});
        } else {
            totalTime += secondMethod;
            sequence.push_back({fastest, slowest});
            sequence.push_back({fastest});
            sequence.push_back({fastest, secondFastest});
            sequence.push_back({fastest});
        }
        people_left.erase(find(people_left.begin(), people_left.end(), slowest));
        people_left.erase(find(people_left.begin(), people_left.end(), secondSlowest));
    }
    if (people_left.size() == 3) {
        sequence.push_back({people_left[0], people_left[1]});
        sequence.push_back({people_left[0]});
        sequence.push_back({people_left[0], people_left[2]});
        people_left.clear();
    }
    if (people_left.size() == 2) {
        sequence.push_back(people_left);
        people_left.clear();
    }
    if (people_left.size() == 1) {
        sequence.push_back(people_left);
        people_left.clear();
    }
    cout << totalTime << endl;
    for (auto i: sequence) {
        for (auto j: i) {
            cout << j;
        }
        cout << endl;
    }
}